scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,266 @@
+use std::cmp::{max, min};
+use std::error::Error;
+use std::time;
+use std::time::SystemTime;
+
+use chrono;
+use futures::stream::StreamExt;
+use scylla::frame::value::Timestamp;
+use scylla::{Session, SessionBuilder};
+use tokio::time::sleep;
+
+type StreamID = Vec<u8>;
+
+pub struct Reader {
+    stream_id_vec: Vec<StreamID>,
+    lower_timestamp: chrono::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+    upper_timestamp: Option<chrono::Duration>,
+    sleep_interval: time::Duration,
+}
+
+impl Reader {
+    pub fn new(
+        stream_ids: Vec<StreamID>,
+        start_timestamp: chrono::Duration,
+        window_size: chrono::Duration,
+        safety_interval: chrono::Duration,
+        sleep_interval: time::Duration,
+    ) -> Reader {
+        Reader {
+            stream_id_vec: stream_ids,
+            lower_timestamp: start_timestamp,
+            window_size,
+            safety_interval,
+            upper_timestamp: std::option::Option::None,
+            sleep_interval,
+        }
+    }
+
+    // Mutable access.
+    fn lower_timestamp_mut(&mut self) -> Option<chrono::Duration> {
+        self.upper_timestamp
+    }
+
+    pub fn set_lower_timestamp(&mut self, new_lower_timestamp: chrono::Duration) {
+        self.upper_timestamp = Some(new_lower_timestamp);
+    }
+
+    pub async fn fetch_cdc(
+        &self,
+        keyspace: String,
+        table_name: &str,
+    ) -> Result<(), Box<dyn Error>> {
+        const URI: &str = "172.17.0.2:9042";
```

Comment:
This must not be fixed like this. Your tests pass just because incidentally your test database's URI matches this one.
This address should be passed to `Reader` through `new` or even better - entire `Session` object should be dependency injected.